doc/mtree_spec_diff.3
.\"
.\" Copyright (c) 2015 Michal Ratajsky <michal@FreeBSD.org>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd August 10, 2015
.Dt MTREE_SPEC_DIFF 3
.Os
.Sh NAME
.Nm mtree_spec_diff
.Nd compare mtree specifications
.Sh LIBRARY
libmtree
.Sh SYNOPSIS
.In mtree.h
.Ft struct mtree_spec_diff *
.Fn mtree_spec_diff_create "struct mtree_spec *spec1" "struct mtree_spec *spec2" "uint64_t keywords" "int options"
.Ft struct mtree_entry *
.Fn mtree_spec_diff_get_spec1_only "struct mtree_spec_diff *sd"
.Ft struct mtree_entry *
.Fn mtree_spec_diff_get_spec2_only "struct mtree_spec_diff *sd"
.Ft struct mtree_entry *
.Fn mtree_spec_diff_get_matching "struct mtree_spec_diff *sd"
.Ft struct mtree_entry *
.Fn mtree_spec_diff_get_different "struct mtree_spec_diff *sd"
.Ft int
.Fn mtree_spec_diff_write_writer "struct mtree_spec_diff *sd" "mtree_writer_fn f" "void *user_data"
.Ft void
.Fn mtree_spec_diff_free "struct mtree_spec_diff *sd"
.In mtree.h
.In mtree_file.h
.Ft int
.Fn mtree_spec_diff_write_file "struct mtree_spec_diff *sd" "FILE *fp"
.Ft int
.Fn mtree_spec_diff_write_fd "struct mtree_spec_diff *sd" "int fd"
.Sh DESCRIPTION
The
.Fn mtree_spec_diff_create
function compares
.Fa spec1
with
.Fa spec2
and creates four lists with information about the differences between entries
in the given specifications. The function also accepts a bitwise OR of
.Fa keywords ,
which allows to restrict the comparison to the given set of keywords. The
.Fa options
argument accepts either a zero or the option:
.Pp
.Bl -tag -offset indent
.It MTREE_SPEC_DIFF_MATCH_EXTRA_KEYWORDS
This option specifies that two entries, which contain different sets of
keywords, but the keywords they have in common have the same values, are considered
matching.
.Pp
By default, without specifying this option, each two keywords with different
sets of keyword are considered to be different.
.El
.Pp
The following functions provide access to the aforementioned lists:
.Pp
.Bl -tag -offset indent
.It Fn mtree_spec_diff_get_spec1_only
Get the
.Em spec1-only
list. This list includes the entries that are present in the first specification, and the
second specification doesn't contain any entry with the same path.
.It Fn mtree_spec_diff_get_spec2_only
Get the
.Em spec2-only
list. This list includes the entries, that are present in the second specification, and the
first specification doesn't contain any entry with the same path.
.It Fn mtree_spec_diff_get_matching
Get the
.Em matching
list. This list includes the entries, that are present in both specifications and have
the same set of keywords with the same keyword values.
.Pp
This list contains pairs of entries from both specs. That means the first entry
in the list comes from
.Fa spec1 ,
and the second entry in the list is the matching entry from
.Fa spec2 .
.Pp
If the
.Em MTREE_SPEC_DIFF_MATCH_EXTRA_KEYWORDS
option is specified, it will also include entries that have the same values in
their common set of keywords, but one entry may contain extra keywords that the
other one doesn't.
.It Fn mtree_spec_diff_get_different
Get the
.Em different
list. This list includes the entries, that have the same paths, but different
keyword values.
.Pp
Just like the
.Em matching
list, the entries in this list are stored in pairs, first entry being from the
first list, second entry being from the second list.
.Pp
Addionally, the entries in this list have their keyword sets modified to only
include the keywords that have different values, or that are not present in
the other spec.
.El
.Pp
All the lists contain copies of entries and any of the lists may be empty.
.Pp
The following functions are provided for writing a report:
.Pp
.Bl -tag -offset indent
.It Fn mtree_spec_diff_write_file
Write the report to the given FILE pointer.
.It Fn mtree_spec_diff_write_fd
Write the report to the given file descriptor.
.It Fn mtree_spec_diff_write_writer
Write the report to the given user-defined function.
.El
.Pp
In the report, as written by these functions, the
.Em spec1-only
entries are prefixed with zero tabs, the
.Em spec2-only
entries with one tab, and the
.Em different
entries with two tabs. Each entry in the
.Em different
column occupies two lines, one from each spec.
.Pp
The
.Fn mtree_spec_diff_free
function releases all allocated resources for the structure, along with the
entry lists.
.Sh RETURN VALUE
The
.Fn mtree_spec_diff_create
function returns a pointer to a newly allocated
.Tn mtree_spec_diff
structure, or NULL if memory couldn't be allocated.
.Pp
The
.Fn mtree_spec_diff_get_spec1_only ,
.Fn mtree_spec_diff_get_spec2_only ,
.Fn mtree_spec_diff_get_matching
and
.Fn mtree_spec_diff_get_different
functions return a pointer to the head of a list, which may be the value
.Dv NULL
when the list is empty.
.Pp
The writing functions return zero on success. On error, they return -1
and set errno to indicate the error.
.Sh SEE ALSO
.Xr mtree 5 ,
.Xr mtree_entry 3 ,
.Xr mtree_spec 3
.Sh AUTHORS
.An -nosplit
The
.Nm libmtree
library was written by
.An Michal Ratajsky Aq michal@FreeBSD.org .


libmtree/mtree_spec_diff.c
/*-
 * Copyright (c) 2015 Michal Ratajsky <michal@FreeBSD.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHORS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <assert.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "mtree.h"
#include "mtree_file.h"
#include "mtree_private.h"

/*
 * Compare spec1 and spec2 entries, producing lists of spec1-only, spec2-only,
 * matching and different entries.
 *
 * Only selected keywords are compared and options allow to specify whether
 * matching and/or different entries should be collected.
 */
struct mtree_spec_diff *
mtree_spec_diff_create(struct mtree_spec *spec1, struct mtree_spec *spec2,
    uint64_t keywords, int options)
{
	struct mtree_spec_diff	*sd;
	struct mtree_entry	*e1, *e2;
	struct mtree_entry	*prev;
	struct mtree_trie	*s2trie;
	uint64_t		 kdiff;
	uint64_t		 kcmp;

	assert(spec1 != NULL);
	assert(spec2 != NULL);

	sd = calloc(1, sizeof(struct mtree_spec_diff));
	if (sd == NULL)
		return (NULL);
	sd->writer = mtree_writer_create();
	if (sd->writer == NULL) {
		free(sd);
		return (NULL);
	}
	s2trie = NULL;
	if (spec1->entries != NULL) {
		sd->s1only = mtree_entry_copy_all(spec1->entries);
		if (sd->s1only == NULL)
			goto err;
	}
	if (spec2->entries != NULL) {
		sd->s2only = mtree_entry_copy_all(spec2->entries);
		if (sd->s2only == NULL)
			goto err;
	}
	/*
	 * If one of the specs has no entries, it's enough to pupulate the
	 * "only" list of the other spec.
	 */
	if (sd->s1only == NULL || sd->s2only == NULL)
		return (sd);

	/*
	 * Convert the 2nd spec to a trie to avoid O^2 comparison.
	 *
	 * This has a side-effect of "merging" the list by replacing former
	 * entries with later ones with the same path.
	 */
	s2trie = mtree_trie_create(NULL);
	if (s2trie == NULL)
		goto err;
	e2 = sd->s2only;
	while (e2 != NULL) {
		if (mtree_trie_insert(s2trie, e2->path, e2) == -1)
			goto err;
		e2 = e2->next;
	}

	/*
	 * Read the 1st list from the end back to the start and for each entry
	 * lookup the path in the 2nd list's trie.
	 *
	 * Reading from the end allows prepending matching entries to either the
	 * match or diff list without having to reverse the lists later.
	 */
	e1 = mtree_entry_get_last(sd->s1only);
	while (e1 != NULL) {
		prev = e1->prev;
		e2 = mtree_trie_find(s2trie, e1->path);
		if (e2 != NULL) {
			/*
			 * A matching entry has been found in the 2nd list,
			 * remove both entries from their lists, compare them and
			 * add them to either match or diff list.
			 *
			 * The remaining entries in s1only will be the ones only
			 * present in the 1st list and the ones in s2only only
			 * in the 2nd list.
			 */
			sd->s1only = mtree_entry_unlink(sd->s1only, e1);
			sd->s2only = mtree_entry_unlink(sd->s2only, e2);

			if (options & MTREE_SPEC_DIFF_MATCH_EXTRA_KEYWORDS) {
				/*
				 * In this case only compare what they have
				 * in common.
				 */
				kcmp = e1->data.keywords & e2->data.keywords;
			} else
				kcmp = keywords;

			if (kcmp == 0 ||
			    mtree_entry_compare_keywords(e1, e2, kcmp, &kdiff) == 0) {
			    	/* Matching. */
				sd->match = mtree_entry_prepend(sd->match, e2);
				sd->match = mtree_entry_prepend(sd->match, e1);
			} else {
				/* Different. */
				mtree_entry_set_keywords(e1,
				    e1->data.keywords & kdiff,
				    MTREE_ENTRY_REMOVE_EXCLUDED);
				mtree_entry_set_keywords(e2,
				    e2->data.keywords & kdiff,
				    MTREE_ENTRY_REMOVE_EXCLUDED);
				sd->diff = mtree_entry_prepend(sd->diff, e2);
				sd->diff = mtree_entry_prepend(sd->diff, e1);
			}
		}
		e1 = prev;
	}
	mtree_trie_free(s2trie);
	return (sd);
err:
	if (s2trie != NULL)
		mtree_trie_free(s2trie);

	mtree_spec_diff_free(sd);
	return (NULL);
}

/*
 * Free the given mtree_spec_diff.
 */
void
mtree_spec_diff_free(struct mtree_spec_diff *sd)
{

	assert(sd != NULL);

	mtree_writer_free(sd->writer);
	mtree_entry_free_all(sd->s1only);
	mtree_entry_free_all(sd->s2only);
	mtree_entry_free_all(sd->match);
	mtree_entry_free_all(sd->diff);
	free(sd);
}

/*
 * Return entries that are only present in the first spec.
 */
struct mtree_entry *
mtree_spec_diff_get_spec1_only(struct mtree_spec_diff *sd)
{

	assert(sd != NULL);

	return (sd->s1only);
}

/*
 * Return entries that are only present in the second spec.
 */
struct mtree_entry *
mtree_spec_diff_get_spec2_only(struct mtree_spec_diff *sd)
{

	assert(sd != NULL);

	return (sd->s2only);
}

/*
 * Return entries that are present in both specs and their keywords match.
 */
struct mtree_entry *
mtree_spec_diff_get_matching(struct mtree_spec_diff *sd)
{

	assert(sd != NULL);

	return (sd->match);
}

/*
 * Return entries that are present in both specs, but their keywords
 * don't match.
 */
struct mtree_entry *
mtree_spec_diff_get_different(struct mtree_spec_diff *sd)
{

	assert(sd != NULL);

	return (sd->diff);
}

/*
 * Write the spec diff in the original mtree's comm(1)-like format.
 */
static int
write_diff(struct mtree_spec_diff *sd)
{
	int ret;

	/*
	 * Only in first spec.
	 */
	mtree_writer_set_format(sd->writer, MTREE_FORMAT_DIFF_FIRST);
	ret = mtree_writer_write_entries(sd->writer, sd->s1only);
	if (ret == -1)
		return (-1);
	/*
	 * Only in second spec.
	 */
	mtree_writer_set_format(sd->writer, MTREE_FORMAT_DIFF_SECOND);
	ret = mtree_writer_write_entries(sd->writer, sd->s2only);
	if (ret == -1)
		return (-1);
	/*
	 * Different in each spec.
	 */
	mtree_writer_set_format(sd->writer, MTREE_FORMAT_DIFF_DIFFER);
	ret = mtree_writer_write_entries(sd->writer, sd->diff);
	if (ret == -1)
		return (-1);

	return (ret);
}

/*
 * Write spec diff to a file.
 */
int
mtree_spec_diff_write_file(struct mtree_spec_diff *sd, FILE *fp)
{

	assert(sd != NULL);
	assert(fp != NULL);

	mtree_writer_set_output_file(sd->writer, fp);

	return (write_diff(sd));
}

/*
 * Write spec diff to a file descriptor.
 */
int
mtree_spec_diff_write_fd(struct mtree_spec_diff *sd, int fd)
{

	assert(sd != NULL);
	assert(fd != -1);

	mtree_writer_set_output_fd(sd->writer, fd);

	return (write_diff(sd));
}

/*
 * Write spec diff to a custom writer.
 */
int
mtree_spec_diff_write_writer(struct mtree_spec_diff *sd, mtree_writer_fn f,
    void *user_data)
{

	assert(sd != NULL);
	assert(f != NULL);

	mtree_writer_set_output_writer(sd->writer, f, user_data);

	return (write_diff(sd));
}
